### ISSUE TITLE
Make value parameter of constant moose objects a controllable parameter

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#6583]
### Description of the enhancement or error report
Constant parameter of 
- [ ] `ConstantAux`
- [ ] `ConstantScalarAux`
- [ ] `ConstantDamper`
- [ ] `ConstantFunction`
- [ ] `ConstantPointSource`
- [ ] `ConstantIC`
- [ ] `GenericConstantMaterial`
- [ ] `ConstantRate`
- [ ] `ConstantDT`
and etc. can be controllable.

### Rationale for the enhancement or information for reproducing the error
I will be needing the `value` parameter in `ConstantFunction` to be controllable but thought maybe it is a good idea to make the similar parameter in the above mentioned classes also controllable. If you especially @aeslaughter disagree, I can change this issue for just changing `ConstantFunction`.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Nobody is using this. We will need to add test though.

**Comment by friedmud** [comment]
If it's not Constant, why would you use one of these objects?

**Comment by permcody** [comment]
I'm not sure I like the idea of making all of our constant objects "not constant".

**Comment by dschwen** [comment]
It's called mutable :-D

**Comment by permcody** [comment]
Yes but I used "not constant" for literary effect!

**Comment by andrsd** [comment]
We use this approach in RELAP-7 quite a bit. Users specify a const value in their input files. But then the python script can change it over time. Just FYI.

On the other hand, I do not see much use in modifying `ConstantIC`. What is the use case @YaqiWang ?

Also, I am not sure about `ConstantDT`. I think if one needs to change dt, they should have a special time stepper. But I'd like to hear the details or motivation behind it...

**Comment by YaqiWang** [comment]
@andrsd I am doing the similar thing. I want the executioner to control the values of the function or aux kernel in order to make the eigenvalue of the system equal to a desired value. A control will link the executioner and the parameter together (That is my plan, not implemented yet). I am not using `ConstantIC`, but thought some people might want to control it. You are probably right about `ConstantDT`. I did grep const in MOOSE and then put the results in this issue, and did not think too much of the usage.

**Comment by andrsd** [comment]
In RELAP, we control the time step via executioner (the old design), but it will be a time stepper in the end (when we fix the design).

> I did grep const in MOOSE and then put the results in this issue, and did not think too much of the usage.

Ok ;-)

Also, `ConstantAux` should be already controllable, because we use that in RELAP. Just FYI. Unless something else is needed with the Control-system @aeslaughter put in.

**Comment by aeslaughter** [comment]
First, at least one of these objects already lies: ConstantDT is anything
but constant.


On Tue, Mar 15, 2016 at 9:18 AM, David Andrs <notifications@github.com>
wrote:

> In RELAP, we control the time step via executioner (the old design), but
> it will be a time stepper in the end (when we fix the design).
>
> I did grep const in MOOSE and then put the results in this issue, and did
> not think too much of the usage.
>
> Ok ;-)
>
> Also, ConstantAux should be already controllable, because we use that in
> RELAP. Just FYI. Unless something else is needed with the Control-system
> @aeslaughter
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_aeslaughter&d=BQMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=h7heP8xwI1i_HikChvhFbEBurKirgfOCdwgBxB9lM8c&m=D9XoNs3bxGZC2rkr5qV4HMV3oEzkeWiGipMCoaGYHEk&s=7mttpl2GKZKHLv2N88vJZcT1CEm6pv1eTmBwzqoirPQ&e=>
> put in.
>
The RELAP7 and MOOSE control logic are still completely separate, so there
is nothing in MOOSE allowing ConstantAux to be controlled currently.


Now I will open a can of worms, what if we allow the user to specify in the
input file which parameters are controllable? I am not sure how a feel
about this and probably won't until I start migrating RELAP7 to the new
system, but just throwing it out there.

[AuxKernels]
  [./const]
    type = ConstantAux
    declare_controllable = 'value'
    value = 1 # this will be a lie, but at least the input file shows us
and the user is aware.
  [../]
[]

**Comment by andrsd** [comment]
@aeslaughter MOOSE objects still need to have `const & TYPE _value`, so they use the value from `InputParams` which we can change via the control system, right?  RELAP is messing with InputParams directly, not via your system. That is the only change I can think of, unless there is more...

**Comment by aeslaughter** [comment]
@andrsd You are correct, RELAP7 is messing with the parameters stored in the warehouse, just not using the Control object. RELAP7 is special and is a `friend` so the non-const version of `getInputParameters` method in `InputParameterWarehouse` is available to the RELAP7 controls.

**Comment by YaqiWang** [comment]
https://github.com/idaholab/moose/blob/devel/framework/src/auxkernels/ConstantAux.C
The value parameter is not controllable.

Another question: should Controls be executed before UO+Aux in `FEProblem::execute()`?

**Comment by andrsd** [comment]
> The value parameter is not controllable.

I understand, that [this](https://github.com/idaholab/moose/blob/devel/framework/include/auxkernels/ConstantAux.h#L46) is the important part that makes it controllable.

**Comment by aeslaughter** [comment]
On Tue, Mar 15, 2016 at 12:48 PM, David Andrs <notifications@github.com>
wrote:

> The value parameter is not controllable.
>
> I understand, that this
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_blob_devel_framework_include_auxkernels_ConstantAux.h-23L46&d=BQMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=h7heP8xwI1i_HikChvhFbEBurKirgfOCdwgBxB9lM8c&m=_F31BhjlvLakp7K7FwUIDZuHBYpQQ6vl8rioPeYciwk&s=B6UojKnjOhl5I3F9dZVAci4Twc0lKUVod8o2eIqf6sI&e=>
> is the important part that makes it controllable.
>
Yes.

**Comment by YaqiWang** [comment]
[This](http://mooseframework.org/wiki/MooseSystems/Controls/) says `declareControllable()`  is a must...

**Comment by andrsd** [comment]
I see. We will have to do that for RELAP-7 update...
